### PR TITLE
PARSER: Extensibility

### PR DESCRIPTION
**Pull request by mikand**
- Refactoring of the parser to allow extensibility
- a few minor hacks to gain performance (~2 seconds executing all the tests in tests/smtlib)

### CODE REVIEW

File: pysmt/smtlib/parser.py [hunk 1]
```diff
@@ -28,7 +29,7 @@
 from pysmt.exceptions import UnknownSmtLibCommandError
 from pysmt.smtlib.script import SmtLibCommand, SmtLibScript
 from pysmt.smtlib.annotations import Annotations
-import pysmt.utils as utils
+from pysmt.utils import interactive_char_iterator
```
Comment (marcogario) [inline pysmt/smtlib/parser.py]: ```interactive_char_iterator``` is only used in the parser. It makes sense to put it here, instead of in ```utils```

File: pysmt/smtlib/parser.py [hunk 2]
```diff
@@ -123,153 +125,159 @@ def unbind_all(self, values):
...
+    else:
+        reader = interactive_char_iterator(handle)
+    c = next(reader)
+
+    eof = False
+    while not eof:
+        if c in specials:
+            # consume all the spaces
+            if c in spaces:
+                while c and c in spaces:
```
Comment (marcogario) [inline pysmt/smtlib/parser.py]: This loop can be moved at the beginning of the outer ```while```, right?

```
while not eof:
    while c and c in spaces:
         c = next(reader)
```

If this is the case, we can probably replace this with an ```ifilter```

File: pysmt/smtlib/parser.py [hunk 3]
```diff
@@ -392,7 +400,7 @@ def _get_basic_type(self, type_name, params=None):
     def _get_var(self, name, type_name, params=None):
         """Returns the PySMT variable corresponding to a declaration"""
         typename = self._get_basic_type(type_name, params)
-        return self._current_env.formula_manager.Symbol(name=name,
+        return self.env.formula_manager.Symbol(name=name,
```
Comment (marcogario) [inline pysmt/smtlib/parser.py]: Maybe we should use ```FormulaManager.get_symbol()```. The difference being that ```get_symbol()``` will return ```None``` if the symbol is not defined, while ```Symbol``` will declare it.

File: pysmt/smtlib/parser.py [hunk 4]
```diff
@@ -509,99 +490,120 @@ def _handle_annotation(self, pyterm, *attrs):
...
+            typename = self.parse_type(tokens, "expression")
+
+            var = self._get_var(vname, typename)
+            self.cache.bind(vname, var)
+            vrs.append(var)
+
+            self.consume_closing(tokens, "expression")
+            current = next(tokens)
+
+
```
Comment (marcogario) [inline pysmt/smtlib/parser.py]: remove \n

File: pysmt/typing.py [hunk 5]
```diff
@@ -65,7 +65,9 @@ def __eq__(self, other):
         return self.type_id == other.type_id
 
     def __ne__(self, other):
-        return not (self == other)
```
Comment (marcogario) [inline pysmt/typing.py]: Why this change? In general, this is not good, since we now require that any sub-class that modifies ```__eq__``` also  modifies ```__ne__```. See ```_BVType```.

**Comment by marcogario** [comment]
Lets add a test in which we extend the parser with some new key-word and function.

File: pysmt/smtlib/parser.py (hunk 3 above)
Comment (mikand) [inline pysmt/smtlib/parser.py]: This is used when a declare-fun is found, hence we really want to obtain the symbol

File: pysmt/typing.py (hunk 5 above)
Comment (mikand) [inline pysmt/typing.py]: The profiler shows this is way more efficient in this way...

File: pysmt/smtlib/parser.py (hunk 4 above)
Comment (mikand) [inline pysmt/smtlib/parser.py]: OK

File: pysmt/smtlib/parser.py (hunk 2 above)
Comment (mikand) [inline pysmt/smtlib/parser.py]: OK

File: pysmt/smtlib/parser.py (hunk 1 above)
Comment (mikand) [inline pysmt/smtlib/parser.py]: I think this is a general-purpose function, do we really want it here?

Comment (marcogario) [inline pysmt/smtlib/parser.py]: I think so, since we only use it in the context of parsing.

File: pysmt/smtlib/parser.py (hunk 3 above)
Comment (marcogario) [inline pysmt/smtlib/parser.py]: ok.

File: pysmt/typing.py (hunk 5 above)
Comment (marcogario) [inline pysmt/typing.py]: Ok, lets fix ```_BVType``` then.
